### PR TITLE
parser: simplify directory ownership semantics

### LINKED ISSUE #32401: Should modules with path attribute become directory owners automatically?
Consider this layout

```
// main.rs
mod foo;
fn main() {}

// foo.rs
mod bar;

// bar.rs
// empty
```

It  generates `cannot declare a new module at this location` error for `mod bar;`, which is good.

However, If I explicitly specify `#[path = ]` for `mod foo;` it compiles:

```
// main.rs
#[path = "foo.rs"]
mod foo;
fn main() {}

// foo.rs
mod bar;

// bar.rs
// empty
```

I don't know if this is expected behavior or not, but it looks suspicious to me .

### PR DESCRIPTION
**Pull request by jseyfried**
This PR simplifies the semantics of "directory ownership". After this PR,
 - an "out of line" module without a `#[path]` attribute (e.g. `mod foo;`) is allowed iff its parent module/block (whichever is nearer) is a directory owner,
 - an out of line module is a directory owner iff its corresponding file is named `mod.rs` (c.f. [comment](https://github.com/rust-lang/rust/issues/32401#issuecomment-201021902)),
 - a block is never a directory owner (c.f. #31534), and
 - an inline module is a directory owner iff either
   - its parent module/block is a directory owner (again, c.f. #31534), or
   - it has a `#[path]` attribute (c.f. #36789).

These semantics differ from today's in three orthogonal ways:
 - `#[path = "foo.rs"] mod foo;` is no longer a directory owner. This is a [breaking-change].
 - #36789 is generalized to apply to modules that are not directory owners in addition to blocks.
 - A macro-expanded out of line module is only allowed where an ordinary out of line module would be allowed. Today, we incorrectly allow macro-expanded out of line modules in modules that are not directory owners (but not in blocks). This is a [breaking-change].

Fixes #32401.
r? @nikomatsakis

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @rust-lang/lang @matklad

**Comment by nikomatsakis** [comment]
Started a crater run.

**Comment by nikomatsakis** [comment]
Results: https://gist.github.com/nikomatsakis/ad150e91da19d7a86d8501a11f2b0a0c

- 6760 crates tested: 4753 working / 1638 broken / **25 regressed** / 2 fixed / 342 unknown.

To be investigated:

* [ ] [petgraph-0.4.1](https://crates.io/crates/petgraph) ([before](https://tools.taskcluster.net/task-inspector/#p5vlI8CJT8ugFHC17pDaIg)) ([after](https://tools.taskcluster.net/task-inspector/#Dm9uqlvTQKyieNz5o7zenQ))
* [ ] [ring-0.5.3](https://crates.io/crates/ring) ([before](https://tools.taskcluster.net/task-inspector/#890rm-mkR6StS2FD69pocA)) ([after](https://tools.taskcluster.net/task-inspector/#jtwIVW-ISKaifB29xre30A))
* [ ] [gtld-data-0.4.1](https://crates.io/crates/gtld-data) ([before](https://tools.taskcluster.net/task-inspector/#RmdtRWX8SiaRefvYzJ_bOw)) ([after](https://tools.taskcluster.net/task-inspector/#lld6Rz08Q4OsvwOWXGzO3Q))
* [ ] [imgur-0.5.0](https://crates.io/crates/imgur) ([before](https://tools.taskcluster.net/task-inspector/#QWGVK34TRQmQkErFMWnX6Q)) ([after](https://tools.taskcluster.net/task-inspector/#lN9-7iUrS--R5PP2YlIWDw))
* [ ] [atomic-0.3.3](https://crates.io/crates/atomic) ([before](https://tools.taskcluster.net/task-inspector/#rBfBzscCTj-dYJbdemCUNQ)) ([after](https://tools.taskcluster.net/task-inspector/#tHWiYQYrQ3WM0M1rtiqi4Q))
* [ ] [presto-0.1.0](https://crates.io/crates/presto) ([before](https://tools.taskcluster.net/task-inspector/#og46wXkPQ5SS17abwRTPeg)) ([after](https://tools.taskcluster.net/task-inspector/#Bakl-LBaSOWpbWiFRG1LZg))
* [ ] [github-0.1.2](https://crates.io/crates/github) ([before](https://tools.taskcluster.net/task-inspector/#LX7TnCVuQme_eomIWTBcwg)) ([after](https://tools.taskcluster.net/task-inspector/#SiWFJkOYR12jrhbUd36nTw))
* [ ] [cargo-deadlinks-0.1.0](https://crates.io/crates/cargo-deadlinks) ([before](https://tools.taskcluster.net/task-inspector/#pLTj6FOUTAmfKu9LNkibVQ)) ([after](https://tools.taskcluster.net/task-inspector/#cx8zl7j0TsSfrGI89Aa3Xw))

**Comment by nikomatsakis** [comment]
@jseyfried so at least a few of those are real regr. We have to decide now: to warning period or what.

cc @rust-lang/lang -- this PR makes some changes to rationalize various aspects of when a module "has a directory" and not (i.e., when you can do `mod foo;`). It breaks some libraries. What to do?

**Comment by brson** [comment]
@nikomatsakis the openssl errors are https://github.com/rust-lang/cargo/issues/3268

**Comment by aturon** [comment]
@nikomatsakis This looks like a good case for applying the general guidelines. In general I'm :+1: going forward with these changes. How hard is a warning period to do in this case? If it's relatively easy, we should do it.

**Comment by brson** [comment]
I'd encourage you to make any breakage here a transitionary warning. We must continue to do our utmost to reduce breakage pain for users. Small breakage adds up.

**Comment by jseyfried** [comment]
It looks like all the breakage is due to `#[path = "foo.rs"] mod foo;` no longer being a directory owner (i.e. https://github.com/rust-lang/rust/issues/32401#issuecomment-201021902). A warning cycle for this is straightforward to implement.

**Comment by nikomatsakis** [comment]
Definitely we should do a warning period.

**Comment by jseyfried** [comment]
I started a warning cycle in the above commit. Since it's not easy to lint from the parser, this is just an ordinary warning with a note about future compatibility.

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -5394,13 +5397,19 @@ impl<'a> Parser<'a> {
                           &format!("maybe move this module `{0}` to its own directory \
                                      via `{0}/mod.rs`",
                                     this_module));
-            if paths.path_exists {
+            if legacy_warn {
+                err.note(
+                    "this was erroneously allowed and will become a hard error in a future release"
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: Hmm. I think we could do an actual lint here -- can't we leave a little note on the AST or something and then issue the lint later, in the lint pass?

At minimum, though, can you open a "forward compat" issue with the appropriate labels and structure and mention it in this message? Roughly follow [the protocol described in RFC 1589](https://github.com/rust-lang/rfcs/blob/master/text/1589-rustc-bug-fix-procedure.md)

Comment (jseyfried) [inline src/libsyntax/parse/parser.rs]: Sure, I'll make it a real lint.

**Comment by nikomatsakis** [comment]
ping @jseyfried -- still planning to make it a real lint?

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (jseyfried) [inline src/libsyntax/parse/parser.rs]: Done.

**Comment by jseyfried** [comment]
@nikomatsakis just fininshed (got sidetracked by some higher-priority stuff)

**Comment by petrochenkov** [comment]
The lint is temporary, but still could you rename it to follow [lint naming conventions](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints)?
Names are supposed to look good in `allow`/`deny` attributes, like `deny(deprecated_directory_ownership)`, but not `deny(directory_ownership)`.

**Comment by jseyfried** [comment]
@petrochenkov yeah, renamed to `legacy_directory_ownership`.

**Comment by nikomatsakis** [comment]
@jseyfried r=me after rebase

**Comment by briansmith** [comment]
Sorry, I don't understand what I need to do to change in *ring* here. I don't want to name all the modules' files mod.rs because it's ridiculous to have dozens of files with the same name in one project. Of course some of them have submodules. What's the alternative?

**Comment by eddyb** [comment]
@briansmith You can declare the module hierarchy recursively, even the entire crate's modules in `lib.rs`, and the nesting matches the directory structure. Some (older?) rustc crates do this, if you need examples.
I believe you can also use `#[path = "foo.rs"] mod foo;` to manually enforce your own hierarchy.

**Pull request merged**